Crash: android.os.TransactionTooLargeException

This one started in 11.8 and has been becoming more prominent in 12.0 and 12.1. 
It's currently the most common crash in Fabric.

Unluckily, the crash logs doesn't say much about where it occurs, but Fabric suggests that it is often seen when transferring bitmaps between Activities, or when saving a large amount of state between Activity configuration changes.

We had other occurrences of it in the past: https://github.com/wordpress-mobile/WordPress-Android/issues/5456.

Also, this thread https://github.com/facebook/react-native/issues/19458 suggests that it may be related to React-Native. The trend we see in Fabric (started in 11.8, growth a bit in 11.9, growth a lot in 12.0 and 12.1) seems to be compatible with mobile Gutenberg deployment in the app.

```
android.os.BinderProxy.transactNative (BinderProxy.java)
android.os.Looper.loop (Looper.java:176)
android.app.ActivityThread.main (ActivityThread.java:6635)
java.lang.reflect.Method.invoke (Method.java)
com.android.internal.os.ZygoteInit.main (ZygoteInit.java:823)
```

5c93a36df8b88c2963f28894-fabric

I've managed consistently reproduce the crash by the following steps:
1. Go to the editor
2. Put a huge text (thousands of symbols)
3. Then go to a media picker to pick a photo
4. Boom - TransactionTooLargeException is thrown

https://github.com/wordpress-mobile/WordPress-Android/assets/16563318/a55c3598-6987-40bf-9c04-b733f0d5e6db

Stacktrace:
```
FATAL EXCEPTION: main
Process: com.jetpack.android.prealpha, PID: 25390
java.lang.RuntimeException: android.os.TransactionTooLargeException: data parcel size 1293520 bytes
Bundle stats:
  android:viewHierarchyState [size=1844]
    android:views [size=1796]
  stateKeyEditorSessionData [size=3696]
  SITE [size=3040]
  androidx.lifecycle.BundlableSavedStateRegistry.key [size=1283200]
    android:support:activity-result [size=3236]
      KEY_COMPONENT_ACTIVITY_REGISTERED_KEYS [size=2712]
    android:support:fragments [size=1279588]
      fragment_5d3a3a21-bae2-4483-9f9e-6885cf8e9080 [size=1265784]
        arguments [size=31204]
          param_gutenberg_props_builder [size=29944]
        viewState [size=1202116]
          0xf [size=1528]
          0xb1 [size=1191432]
        childFragmentManager [size=31208]
          fragment_873fae23-e456-4e1a-896c-050a1630f466 [size=30788]
            arguments [size=30024]
              param_gutenberg_props_builder [size=29944]
      fragment_5c4703a2-afc6-4db0-af00-2d1359b4e04b [size=5752]
        viewState [size=4756]
      fragment_79012020-1a29-481a-be30-8a0805b110e5 [size=1436]
      fragment_a991c6d3-da8c-4d39-b612-4fec18a0d14f [size=5428]
        arguments [size=3128]
          key_site [size=3040]
        viewState [size=1548]
PersistableBundle stats:
  [null]
	at android.app.servertransaction.PendingTransactionActions$StopInfo.run(PendingTransactionActions.java:146)
	at android.os.Handler.handleCallback(Handler.java:958)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.app.ActivityThread.main(ActivityThread.java:8177)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
Caused by: android.os.TransactionTooLargeException: data parcel size 1293520 bytes
	at android.os.BinderProxy.transactNative(Native Method)
	at android.os.BinderProxy.transact(BinderProxy.java:584)
	at android.app.IActivityClientController$Stub$Proxy.activityStopped(IActivityClientController.java:1439)
	at android.app.ActivityClient.activityStopped(ActivityClient.java:96)
	at android.app.servertransaction.PendingTransactionActions$StopInfo.run(PendingTransactionActions.java:135)
	at android.os.Handler.handleCallback(Handler.java:958) 
	at android.os.Handler.dispatchMessage(Handler.java:99) 
	at android.os.Looper.loopOnce(Looper.java:205) 
	at android.os.Looper.loop(Looper.java:294) 
	at android.app.ActivityThread.main(ActivityThread.java:8177) 
	at java.lang.reflect.Method.invoke(Native Method) 
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552) 
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971) 
```

--- a/RELEASE-NOTES.txt
+++ b/RELEASE-NOTES.txt
@@ -10,6 +10,7 @@
 * [*] [Jetpack-only] Site Monitoring: Add Metrics, PHP Logs, and Web Server Logs under Site Monitoring [https://github.com/wordpress-mobile/WordPress-Android/issues/20067]
 * [**] Prevent images from temporarily disappearing when uploading media [https://github.com/WordPress/gutenberg/pull/57869]
 * [***] [Jetpack-only] Reader: introduced new UI/UX for content navigation and filtering [https://github.com/wordpress-mobile/WordPress-Android/pull/19978]
+* [**] Prevents crashes when the webview state is too big [https://github.com/wordpress-mobile/WordPress-Android/pull/20139]
 
 24.1
 -----

--- a/WordPress/src/main/java/org/wordpress/android/ui/WebViewActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/WebViewActivity.java
@@ -28,6 +28,9 @@ public abstract class WebViewActivity extends LocaleAwareActivity {
 
     private static final String URL = "url";
 
+    private static final String WEBVIEW_CHROMIUM_STATE = "WEBVIEW_CHROMIUM_STATE";
+    private static final int WEBVIEW_CHROMIUM_STATE_THRESHOLD = 300 * 1024; // 300 KB
+
     protected WebView mWebView;
 
     @Override
@@ -84,8 +87,17 @@ public abstract class WebViewActivity extends LocaleAwareActivity {
      * save the webView state with the bundle so it can be restored
      */
     @Override
-    protected void onSaveInstanceState(Bundle outState) {
+    protected void onSaveInstanceState(@NonNull Bundle outState) {
         mWebView.saveState(outState);
+
+        // If the WebView state is too large, remove it from the bundle. This workaround is
+        // necessary since the Android system cannot handle large states without a crash.
+        // Note that Chromium `WebViewBrowserFragment` uses a similar workaround for this issue:
+        // https://source.chromium.org/chromium/chromium/src/+/27a9bbd3dcd7005ac9f3862dc2e356b557023de9
+        byte[] webViewState = outState.getByteArray(WEBVIEW_CHROMIUM_STATE);
+        if (webViewState != null && webViewState.length > WEBVIEW_CHROMIUM_STATE_THRESHOLD) {
+            outState.remove(WEBVIEW_CHROMIUM_STATE);
+        }
         super.onSaveInstanceState(outState);
     }
 
@@ -93,7 +105,7 @@ public abstract class WebViewActivity extends LocaleAwareActivity {
      * restore the webView state saved above
      */
     @Override
-    protected void onRestoreInstanceState(Bundle savedInstanceState) {
+    protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
         super.onRestoreInstanceState(savedInstanceState);
         mWebView.restoreState(savedInstanceState);
     }

--- a/WordPress/src/main/java/org/wordpress/android/ui/WebViewActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/WebViewActivity.java
@@ -14,8 +14,10 @@ import androidx.appcompat.widget.Toolbar;
 
 import org.wordpress.android.R;
 import org.wordpress.android.WordPress;
+import org.wordpress.android.analytics.AnalyticsTracker;
 import org.wordpress.android.util.extensions.CompatExtensionsKt;
 
+import java.util.HashMap;
 import java.util.Map;
 
 /**
@@ -90,13 +92,18 @@ public abstract class WebViewActivity extends LocaleAwareActivity {
     protected void onSaveInstanceState(@NonNull Bundle outState) {
         mWebView.saveState(outState);
 
-        // If the WebView state is too large, remove it from the bundle. This workaround is
+        // If the WebView state is too large, remove it from the bundle and track the error. This workaround is
         // necessary since the Android system cannot handle large states without a crash.
         // Note that Chromium `WebViewBrowserFragment` uses a similar workaround for this issue:
         // https://source.chromium.org/chromium/chromium/src/+/27a9bbd3dcd7005ac9f3862dc2e356b557023de9
         byte[] webViewState = outState.getByteArray(WEBVIEW_CHROMIUM_STATE);
         if (webViewState != null && webViewState.length > WEBVIEW_CHROMIUM_STATE_THRESHOLD) {
             outState.remove(WEBVIEW_CHROMIUM_STATE);
+
+            // Track the error to better understand the root of the issue
+            Map<String, String> properties = new HashMap<>();
+            properties.put(URL, mWebView.getUrl());
+            AnalyticsTracker.track(AnalyticsTracker.Stat.WEBVIEW_TOO_LARGE_PAYLOAD_ERROR, properties);
         }
         super.onSaveInstanceState(outState);
     }

--- a/WordPress/src/main/java/org/wordpress/android/ui/WebViewActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/WebViewActivity.java
@@ -100,9 +100,13 @@ public abstract class WebViewActivity extends LocaleAwareActivity {
         if (webViewState != null && webViewState.length > WEBVIEW_CHROMIUM_STATE_THRESHOLD) {
             outState.remove(WEBVIEW_CHROMIUM_STATE);
 
+            // Save the URL so it can be restored later
+            String url = mWebView.getUrl();
+            outState.putString(URL, url);
+
             // Track the error to better understand the root of the issue
             Map<String, String> properties = new HashMap<>();
-            properties.put(URL, mWebView.getUrl());
+            properties.put(URL, url);
             AnalyticsTracker.track(AnalyticsTracker.Stat.WEBVIEW_TOO_LARGE_PAYLOAD_ERROR, properties);
         }
         super.onSaveInstanceState(outState);
@@ -114,7 +118,14 @@ public abstract class WebViewActivity extends LocaleAwareActivity {
     @Override
     protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
         super.onRestoreInstanceState(savedInstanceState);
-        mWebView.restoreState(savedInstanceState);
+        if (savedInstanceState.containsKey(WEBVIEW_CHROMIUM_STATE)) {
+            mWebView.restoreState(savedInstanceState);
+        } else {
+            String url = savedInstanceState.getString(URL);
+            if (url != null) {
+                mWebView.loadUrl(url);
+            }
+        }
     }
 
     @Override

--- a/libs/analytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java
+++ b/libs/analytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java
@@ -1106,7 +1106,8 @@ public final class AnalyticsTracker {
         SITE_MONITORING_SCREEN_SHOWN,
         OPENED_SITE_MONITORING,
         SITE_MONITORING_TAB_SHOWN,
-        SITE_MONITORING_TAB_LOADING_ERROR
+        SITE_MONITORING_TAB_LOADING_ERROR,
+        WEBVIEW_TOO_LARGE_PAYLOAD_ERROR,
     }
 
     private static final List<Tracker> TRACKERS = new ArrayList<>();

--- a/libs/analytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java
+++ b/libs/analytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java
@@ -2707,6 +2707,8 @@ public class AnalyticsTrackerNosara extends Tracker {
                 return "site_monitoring_tab_shown";
             case SITE_MONITORING_TAB_LOADING_ERROR:
                 return "site_monitoring_tab_loading_error";
+            case WEBVIEW_TOO_LARGE_PAYLOAD_ERROR:
+                return "webview_too_large_payload_error";
         }
         return null;
     }

